# Step 1: Data health check

In [1]:
import pandas as pd
cc_data = pd.read_csv('data.csv',names=['AccountId','MerchantId','TransactionAmount','Extra'])
cc_data

AccountId   MerchantId TransactionAmount   Extra
0       111184657   6.8579E+13                 1     NaN
1       337612045   5.7186E+13             99.99     NaN
2       267635954   1001515665             30.01     NaN
3        23826603      8146524                 X   14.96
4       300553673      4650693                 X    8.99
5       359135601      7981962             39.99     NaN
6       346875322      4975861                 X   20.00
7       323678754     33898193             347.5     NaN
8       359135601     47337102                30     NaN
9       284170648   5.7186E+13             26.56     NaN
10      240965569     12155272               288     NaN
11      108841143     12155272               828     NaN
12      346879928    9.802E+11                 X    2.95
13       93745072     31183072                 X    7.25
14      110639786     61494462            269.98     NaN
15      273353739     31183072             12.47     NaN
16      335186751  3.45852E+14                 X    4.50
17      337077984      6898266               2.4     NaN
18      358797248     60772753                 X   94.00
19      109606726     31183072             12.59     NaN
20      341119467   5.7183E+13              0.99     NaN
21      259969737   1002217709             60.42     NaN
22      361567428  1.82128E+14              0.65     NaN
23      172589123     63577983                10     NaN
24      111249517   9.8002E+11              4.95     NaN
25      350232814  5.54197E+14              10.8     NaN
26      327440117     63577983                10     NaN
27      303471104     63577983                20     NaN
28      338427166     63577983                25     NaN
29      285642030  5.20335E+14              2.99     NaN
...           ...          ...               ...     ...
249970  328666629  5.54196E+14                 X   14.00
249971  110590718     78418472                24     NaN
249972  304679568  5.20335E+14                 X   13.50
249973  340399709      4319401              3.76     NaN
249974  340387536      1806871             85.16     NaN
249975  354204643      7244908             25.55     NaN
249976  108900786     54186291                89     NaN
249977  195165432   5.7193E+13             53.47     NaN
249978  350295230     27355042                65     NaN
249979  110173734     84437561             10.89     NaN
249980  110020702      4612248             15.98     NaN
249981  109232371     53146663             10.55     NaN
249982  254770503     53254361              1.55     NaN
249983  339153739     42816663               2.4     NaN
249984  200315034   9.8002E+11              2.99     NaN
249985  109266634  5.20335E+14                90     NaN
249986  111307703     58415582             14.75     NaN
249987  363619213      3005572             12.92     NaN
249988  356761960   2100381383              30.9     NaN
249989  349530258     17511333                 9     NaN
249990  108955447      6653981              4.98     NaN
249991  292066507     39226413                 1     NaN
249992  219803020      9980463            451.17     NaN
249993  327974695     55789351              4.74     NaN
249994  354820625     43761443             70.94     NaN
249995  341921146      6026942                 X  250.00
249996  356011511  5.54197E+14           1678.95     NaN
249997  341929230     38028912                 X    6.54
249998  325725933     41430542                20     NaN
249999  230614994      7529431                28     NaN

[250000 rows x 4 columns]

In [2]:
cc_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 250000 entries, 0 to 249999
Data columns (total 4 columns):
AccountId            250000 non-null int64
MerchantId           249477 non-null object
TransactionAmount    248026 non-null object
Extra                46656 non-null float64
dtypes: float64(1), int64(1), object(2)
memory usage: 7.6+ MB


In [3]:
cc_data.isnull().sum()      #checking for missing data

AccountId                 0
MerchantId              523
TransactionAmount      1974
Extra                203344
dtype: int64

In [4]:
cc_data.notnull().sum()      #checking for missing data

AccountId            250000
MerchantId           249477
TransactionAmount    248026
Extra                 46656
dtype: int64

In [5]:
cc_data["TransactionAmount"].value_counts()

X          47763
1           4811
10          2817
20          2654
50          1258
5           1252
30          1212
15          1133
2           1071
40          1033
25          1005
9.99         944
5.99         818
0.64         805
12           794
4.99         762
1.58         757
6            751
0.69         741
0.01         728
18           719
3            715
8            696
0.99         661
60           659
100          642
14.99        627
7.99         626
0.65         626
35           565
           ...  
91.88          1
592.5          1
204.84         1
223.78         1
219.12         1
2063.9         1
106.27         1
182.16         1
214.2          1
97.57          1
1212.15        1
176.04         1
1208           1
120.3          1
305.83         1
479.29         1
1402.5         1
145.76         1
521.81         1
342.97         1
201.96         1
173.98         1
270.44         1
721.05         1
642.68         1
122.38         1
140.52         1
258.01        

# Annalysis Report

#### 1- Dataset consists of 4 columns AccountId, MerchantId, TransactionAmount & Extra.
 
#### Approach 1
 
####  Every 'X' value in 'TransactionAmount' column has a corresponding value in 'Extra' columns (Assuming that it Denotes Data corruption) we can replace every 'X' transaction with it's corresponding value in 'Extra' column to fill up the 'TransactionAmount' column.
####  But further analysis shows count of X=47763 and values in Extra=4656 which does not match thus idea of replacing fails.
  
####  Approach 2
  
####  Ignore the 'Extra' columns and carry on with the analysis by replacing 'X' with 'NaN' i.e No Transaction (on a safer side) rather than replacing it with some mean value which will be easily affected by outliers.
  
  #### moving forward with Approach 2.
  
#### 2- Datatype of the columns needs to be change since it is a transaction amount it can't be an object.( float or int )

#### 3- One AccountId can have transaction with many MerchantId so we create group of every AccountId for further analysis.

# Step 2: Simple statistics

#### To build up knowledge of the typical transaction, you accumulate some simple statistics about each customer. Write a program that reads event data from a file, tracks the mean transaction amount (η) for each customer and its standard deviation (σ), and prints it to standard output after every transaction.

### The following code replaces "X" and changes datatye of TransactionAmount
### Output shows Mean, Count and STD per AccountID

In [6]:
def track_cust():
    
    cc_data = pd.read_csv('data.csv',names=['AccountId','MerchantId','TransactionAmount','Extra'])
    
    cc_data['TransactionAmount'] = cc_data['TransactionAmount'].replace(['X'],'NaN')
    
    cc_data['TransactionAmount'] = cc_data['TransactionAmount'].astype(float)
    
    cc_data_main = cc_data[['AccountId','TransactionAmount']]
    
    print(cc_data_main.groupby(['AccountId']).agg(['count','mean','std']).sort_values(by=('TransactionAmount','count'),ascending = False))

track_cust()

          TransactionAmount                          
                      count         mean          std
AccountId                                            
111232879               150   299.587133   350.337043
341365559               122   172.859180   227.521641
111249517               100    40.627100    42.122955
355162409                94     1.237553     1.086298
285980289                92    26.897283    72.332817
292395225                80     6.376875     7.635337
330376442                80    37.591125   140.486642
109713322                78    52.666795   101.362592
271868304                76    37.157237    21.992372
109307289                74    23.431081    23.848938
110392237                73    27.583562    29.686164
340556454                73    58.774795   185.659986
108913899                73    16.731233    21.523352
283040627                73    35.482603   101.358142
109934363                72    35.066111    43.009497
334993816                72 

####  Suppose you observe a new transaction for d dollars. You can use η and σ to detect whether it’s an anomaly by checking if d differs from η by at least 3 standard deviations. Update your program with this simple detector and output any warnings into a log. Is your detector working reliably when it has seen very few transactions per customer? One approach to mitigate the initial lack of data is to fall back on the global statistics. Implement tracking mean transaction amount and its variance globally (for all customers) and update your detector to use this information when needed. 


# Anomaly Detection model
### Output 1) Transaction 'count', Transaction amount 'Mean' & 'STD' per Account ID
### 2) Detecting anomaly by checking if new transaction differs from η by at least 3 standard deviations
### 3) User Input Account No. and Amount
### 4) User's Transaction History
### 5) Detecting anomaly for Account Id with transacton count < 3 on the basis of global statistics. 

## Step 3 included

In [7]:
import pandas as pd
import numpy as np
import logging

def detect_anomaly():
    
    cc_data = pd.read_csv('data.csv',names=['AccountId','MerchantId','TransactionAmount','Extra'])
    
    cc_data['TransactionAmount'] = cc_data['TransactionAmount'].replace(['X'],'NaN')
    
    cc_data['TransactionAmount'] = cc_data['TransactionAmount'].astype(float)
    
    global_mean = cc_data['TransactionAmount'].mean()
    global_var = cc_data['TransactionAmount'].var()
    global_std = np.sqrt(global_var)
    
    cc_data_main = cc_data[['AccountId','TransactionAmount']]
    
    print(cc_data_main.groupby(['AccountId']).agg(['count','mean','std']).sort_values(by=('TransactionAmount','count'),ascending = False),"\n")
    
    lower_range = cc_data_main.groupby(['AccountId']).mean() - 3 * cc_data_main.groupby(['AccountId']).std()
    upper_range = cc_data_main.groupby(['AccountId']).mean() + 3 * cc_data_main.groupby(['AccountId']).std()
    
    inp_acc = int(input("Enter an Account No : "))
    inp_amt  = float(input("Enter amount : "))
    history = cc_data[(cc_data.AccountId == inp_acc)]
    
    acc_tran_count = float(cc_data_main.groupby(['AccountId']).count().loc[inp_acc])
    
    
    if acc_tran_count > 2 :
        
        print("Lower Range of Transaction amount for Account Id", inp_acc, "is : ", float(lower_range.loc[inp_acc]))
        print("Upper Range of Transaction amount for Account Id", inp_acc, "is : ", float(upper_range.loc[inp_acc]),"\n")
      
        if inp_amt < float(lower_range.loc[inp_acc]) or inp_amt > float(upper_range.loc[inp_acc]):
                                
            logging.warning('Anomaly detected \n')
            print("History of Account Id ",inp_acc, "is \n",history)
                    
        else:
            
            print("Transaction amount acceptable \n")
            print("History of Account Id ",inp_acc, "is \n",history)
            print("Minimum Transaction amount in past: ",history["TransactionAmount"].min())
            print("Maximum Transaction amount in past: ",history["TransactionAmount"].max())
            
    else:
        
        print("Global Mean amount for dataset is : ",global_mean)
        
        if inp_amt > float(global_mean):
            
            logging.warning('Anomaly detected \n')
            print("History of Account Id ",inp_acc, "is \n",history)
            
        else:
            
            print("Transaction amount acceptable \n")
            print("History of Account Id ",inp_acc, "is \n",history)
            
        
detect_anomaly()

          TransactionAmount                          
                      count         mean          std
AccountId                                            
111232879               150   299.587133   350.337043
341365559               122   172.859180   227.521641
111249517               100    40.627100    42.122955
355162409                94     1.237553     1.086298
285980289                92    26.897283    72.332817
292395225                80     6.376875     7.635337
330376442                80    37.591125   140.486642
109713322                78    52.666795   101.362592
271868304                76    37.157237    21.992372
109307289                74    23.431081    23.848938
110392237                73    27.583562    29.686164
340556454                73    58.774795   185.659986
108913899                73    16.731233    21.523352
283040627                73    35.482603   101.358142
109934363                72    35.066111    43.009497
334993816                72 

## Test Case 1
### a) Regular Account Id with Transaction count 100

In [8]:
detect_anomaly()

          TransactionAmount                          
                      count         mean          std
AccountId                                            
111232879               150   299.587133   350.337043
341365559               122   172.859180   227.521641
111249517               100    40.627100    42.122955
355162409                94     1.237553     1.086298
285980289                92    26.897283    72.332817
292395225                80     6.376875     7.635337
330376442                80    37.591125   140.486642
109713322                78    52.666795   101.362592
271868304                76    37.157237    21.992372
109307289                74    23.431081    23.848938
110392237                73    27.583562    29.686164
340556454                73    58.774795   185.659986
108913899                73    16.731233    21.523352
283040627                73    35.482603   101.358142
109934363                72    35.066111    43.009497
334993816                72 

### b) Regular Account Id with Transaction count 100 (Anomaly Detected)

In [9]:
detect_anomaly()

          TransactionAmount                          
                      count         mean          std
AccountId                                            
111232879               150   299.587133   350.337043
341365559               122   172.859180   227.521641
111249517               100    40.627100    42.122955
355162409                94     1.237553     1.086298
285980289                92    26.897283    72.332817
292395225                80     6.376875     7.635337
330376442                80    37.591125   140.486642
109713322                78    52.666795   101.362592
271868304                76    37.157237    21.992372
109307289                74    23.431081    23.848938
110392237                73    27.583562    29.686164
340556454                73    58.774795   185.659986
108913899                73    16.731233    21.523352
283040627                73    35.482603   101.358142
109934363                72    35.066111    43.009497
334993816                72 

Lower Range of Transaction amount for Account Id 111249517 is :  -85.74176356255214
Upper Range of Transaction amount for Account Id 111249517 is :  166.99596356255205 

History of Account Id  111249517 is 
         AccountId   MerchantId  TransactionAmount  Extra
24      111249517   9.8002E+11               4.95    NaN
10122   111249517      3763877              12.30    NaN
11614   111249517      3041597             160.00    NaN
14975   111249517     83574772              60.90    NaN
15648   111249517   6.8579E+13               1.00    NaN
20360   111249517     54782681              16.40    NaN
24777   111249517  5.20335E+14               8.10    NaN
25637   111249517     69827132              17.18    NaN
25879   111249517      8832461              17.67    NaN
39721   111249517     40378212               1.00    NaN
45404   111249517     12570663              50.40    NaN
48674   111249517  2.06996E+14               5.99    NaN
52772   111249517      6870570              16.20  

### Test Case 2
#### a) Account Id with Transaction count < 3 ( On the basis of Global Statistics )

In [10]:
detect_anomaly()

          TransactionAmount                          
                      count         mean          std
AccountId                                            
111232879               150   299.587133   350.337043
341365559               122   172.859180   227.521641
111249517               100    40.627100    42.122955
355162409                94     1.237553     1.086298
285980289                92    26.897283    72.332817
292395225                80     6.376875     7.635337
330376442                80    37.591125   140.486642
109713322                78    52.666795   101.362592
271868304                76    37.157237    21.992372
109307289                74    23.431081    23.848938
110392237                73    27.583562    29.686164
340556454                73    58.774795   185.659986
108913899                73    16.731233    21.523352
283040627                73    35.482603   101.358142
109934363                72    35.066111    43.009497
334993816                72 

#### b) Account Id with Transaction count < 3 (Anomaly Detected On the basis of Global Statistics )

In [11]:
detect_anomaly()

          TransactionAmount                          
                      count         mean          std
AccountId                                            
111232879               150   299.587133   350.337043
341365559               122   172.859180   227.521641
111249517               100    40.627100    42.122955
355162409                94     1.237553     1.086298
285980289                92    26.897283    72.332817
292395225                80     6.376875     7.635337
330376442                80    37.591125   140.486642
109713322                78    52.666795   101.362592
271868304                76    37.157237    21.992372
109307289                74    23.431081    23.848938
110392237                73    27.583562    29.686164
340556454                73    58.774795   185.659986
108913899                73    16.731233    21.523352
283040627                73    35.482603   101.358142
109934363                72    35.066111    43.009497
334993816                72 

Global Mean amount for dataset is :  74.3367538686795
History of Account Id  318393181 is 
         AccountId MerchantId  TransactionAmount  Extra
241799  318393181   48218562                NaN  16.19
246533  318393181    2606010                NaN  41.66
